score-server/src/main/java/bio/overture/score/server/repository/s3/S3DownloadService.java

```diff
@@ -136,21 +112,56 @@ public ObjectSpecification download(
       } else {
         parts = partCalculator.divide(offset, length);
       }
+      if (objectSpec == null) {
+        ObjectMetadata metadata =
+            s3Client.getObjectMetadata(
+                bucketNamingService.getStateBucketName(objectId), objectKey.getKey());
+        fillPartUrls(objectKey, parts, false, forExternalUse);
+        objectSpec =
+            new ObjectSpecification(
+                objectKey.getKey(), objectId, objectId, parts, length, metadata.getETag(), false);
+      }
+
+      // Short-circuit in default case
+      if (objectSpec != null) {
+        if (!forExternalUse && (offset == 0L && length < 0L)) {
+          return excludeUrls ? removeUrls(objectSpec) : objectSpec;
+        }
 
-      fillPartUrls(objectKey, parts, objectSpec.isRelocated(), forExternalUse);
+        // Calculate range values
+        // To retrieve to the end of the file
+        if (!forExternalUse && (length < 0L)) {
+          length = objectSpec.getObjectSize() - offset;
+        }
 
-      val spec =
-          new ObjectSpecification(
-              objectKey.getKey(),
-              objectId,
-              objectId,
-              parts,
-              length,
-              objectSpec.getObjectMd5(),
-              objectSpec.isRelocated());
+        // Validate offset and length parameters:
+        // Check if the offset + length > length - that would be too big
+        if ((offset + length) > objectSpec.getObjectSize()) {
+          throw new InternalUnrecoverableError(
+              "Specified parameters exceed object size (object id: "
+                  + objectId
+                  + ", offset: "
+                  + offset
+                  + ", length: "
+                  + length
+                  + ")");
+        }
+      }
+      if (objectSpec != null) {
```

Comment:
This `if` is duplicate of the same if from line 126. We don't need both of them.